packages/query/src/block-processor.ts

```diff
@@ -133,13 +133,19 @@ export class BlockProcessor implements BlockProcessorInterface {
     const startHeight = fullSyncHeight ? fullSyncHeight + 1n : 0n;
     const latestBlockHeight = await this.querier.tendermint.latestBlockHeight();
 
+    let nextEpochStartHeight: bigint | undefined = startHeight === 0n ? 0n : undefined;
+
```

Comment:
This is probably a very rare case and I won't be able to reproduce it, but if we stop sync after saving the last block of an epoch and then start sync again, we miss an epoch.

Reply:
Damn, you're right. What do you think — maybe I should store `nextEpochStartHeight` in IndexedDB? 😩

Reply:
I had an idea to check the height of the epoch change by condition `height % epochDuration === 0`, but this is probably bad, `epochDuration` can change

Reply:
It would be convenient if the compact block had some indicator that determines the first block of the epoch, not the last one

Reply:
> I had an idea to check the height of the epoch change by condition height % epochDuration === 0, but this is probably bad, epochDuration can change

Yeah, that won't work — an epoch can stop short if e.g., a validator is slashed.

> It would be convenient if the compact block had some indicator that determines the first block of the epoch, not the last one

YEP 😅

Reply:
Just had a thought — I could, here at the end, insert the epoch into the database, since I'll already know it's start height. That'd work, right?

Reply:
> Just had a thought — I could, here at the end, insert the epoch into the database, since I'll already know it's start height. That'd work, right?

Yes

But you must update the validator info in the first epoch block, in any case the block processor must be able to detect the first epoch block

Reply:
Got it. I'll work on this. Thanks for flagging this issue!

Reply:
@Valentine1898 actulaly, I'll do that as part of my next PR where I store validator infos in the database, if that's OK with you (since this is SUCH an edge case). Mind approving this in the meantime?

Reply:
Uh, yeah, sure